sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -104,7 +104,8 @@ case class InMemoryTableScanExec(
 
     case In(a: AttributeReference, list: Seq[Expression]) if list.forall(_.isInstanceOf[Literal]) =>
       list.map(l => statsFor(a).lowerBound <= l.asInstanceOf[Literal] &&
-        l.asInstanceOf[Literal] <= statsFor(a).upperBound).reduce(_ || _)
+        l.asInstanceOf[Literal] <= statsFor(a).upperBound)
```

Comment:
This still looks more complex and less efficient than 

```
list.exists(l => statsFor(a).lowerBound <= l.asInstanceOf[Literal] && l.asInstanceOf[Literal] <= statsFor(a).upperBound)
```

or better

```
val stats = statsFor(a)
list.exists { l =>
  val literal = l.asInstanceOf[Literal]
  stats.lowerBound <= literal && literal <= stats.upperBound
}
```

The point being that you should be able to short-circuit evaluation here.
Or have I missed something basic like that these aren't Booleans?

Reply:
Exactly, they aren't Booleans. This they are `Expression`s. Thus I can't think of a different (more efficient) way of doing this.

Reply:
I see. How does `.contains(true)` work then? or did that not work?
I suppose all I mean is that we should write something that works on an empty list (returns false?) and also short-circuits (stops when anything is true). Is that possible?

Reply:
It was a mistake, sorry. It returned always `false`.
I see what you mean, but in this piece of code we are only building the `Expression` and we are not evaluating it. Thus it is not possible to short-circuit, because the `Expression` must be built entirely.

Reply:
OK I understand. Rather than build an extra term into the expression every time, why not special-case an empty list? return Literal(false) if empty, otherwise return the same thing as before.

Reply:
I think this is a bit more elegant and concise as code style, but your idea can be more efficient (though I am not sure how much overhead is introduced by a `False` evaluation). Then I am updating this PR according to your suggestion, thanks.